### PR TITLE
 Write Unit Tests and Update Query() to pass them.

### PR DESCRIPTION
**Pull request by amuamushu**
As the second part of TDD, I wrote the unit tests for the required scenarios and updated query() to pass the tests.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -22,59 +22,87 @@
...
-    Collection<TimeRange> times = new ArrayList<>();
-    Collection<String> attendees = request.getAttendees();
-    
+    timesForEveryone = new ArrayList<>();
+    timesForConfirmed = new ArrayList<>();
+
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
       // If the event is longer than a day, no time slots are available to book a meeting.
-      return times; 
+      return timesForEveryone;
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: I would return new ArrayList<>(). This makes it clear that nothing is returned.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -22,59 +22,87 @@
...
-        currentEndTime = time.end();
-        continue;
-      }
-      TimeRange gapTimeRange = TimeRange.fromStartEnd(currentEndTime, time.start(), false);
-      currentEndTime = time.end();
-      times.add(gapTimeRange);
+      addTimeGapIfPossible(request, eventAttendees, event);
     }
-
+
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion

```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -22,59 +22,87 @@
...
+      return timesForConfirmed;
+    }
+    return timesForEveryone;
+  }
+
+  /**
+   * Adds a TimeRange to indicate a time gap for either {@code timesForEveryone} or {@code timesForConfirmed} depending
+   * on whether optional or confirmed attendees are attending the current {@code event}.
+   * 
+   * <p>Checks {@code request.duration} and {@code eventAttendees} to see if a time gap is possible for which people.
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: "if a time gap is possible for while people". Not sure I understand this. Can you rephrase?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -22,59 +22,87 @@
...
+   */
+  public void addTimeGapIfPossible(MeetingRequest request, Set<String> eventAttendees, Event event) {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: usually if you are setting a value for a class member you prepend the method name with 'set'. So for this I would rename 'setTimeGapIfPossible'.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -25,7 +25,10 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
-/** */
+/**
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: You do not need a comment for test classes.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -34,14 +37,17 @@
   // Some people that we can use in our tests.
   private static final String PERSON_A = "Person A";
   private static final String PERSON_B = "Person B";
+  private static final String PERSON_C = "Person C";
 
   // All dates are the first day of the year 2020.
   private static final int TIME_0800AM = TimeRange.getTimeInMinutes(8, 0);
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: can you add inline comments for the raw variables? Example:

TimeRange.getTimeInMinutes(/*argumentName1=*/ 8, /*argumentName2=*/0);

This will help me understand what these numbers mean.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 7]
```diff
@@ -200,7 +201,7 @@ public void doubleBookedPeople() {
   }
 
   @Test
-  public void justEnoughRoom() {
+  public void query_justEnoughRoom() {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I am not sure what justEnoughRoom means. Can you use a more accurate wording? Maybe it should be renamed as: query_oneTimeSlotAvailable

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 8]
```diff
@@ -175,7 +177,7 @@ public void nestedEvents() {
   }
 
   @Test
-  public void doubleBookedPeople() {
+  public void query_doubleBookedPeople() {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: usually you title these unit tests as <method>_<conditions>_<expectedResults>. So for this unit test I think it should be named as: query_doubleBookedAttendees_<expectedResult> or is it query_noAvailableSlots_doubleBookedAttendees?

Comment (billie-cat) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I think your other unit tests also need to be renamed to follow this pattern.

**Review by billie-cat — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (f-camp) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: +1

Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: done

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 8 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I updated the method names to include the number of available time slots.
Done!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 7 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I used your suggestion.
Done!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 5 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Got it
Done

**Review by f-camp — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -22,59 +22,87 @@
 
 /** Finds potential meeting times. */
 public final class FindMeetingQuery {
+  public Collection<TimeRange> timesForEveryone;
+  public Collection<TimeRange> timesForConfirmed;
+  public int currentEndTime;
 
   /**
    * Returns available TimeRanges for the {@code request} given {@code events}, a Collection of
    * all events so far.
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The text: a Collection of all events so far can be removed given we know from the method declaration that its a collection.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -22,59 +22,87 @@
...
-    Collection<TimeRange> times = new ArrayList<>();
-    Collection<String> attendees = request.getAttendees();
-    
+    timesForEveryone = new ArrayList<>();
+    timesForConfirmed = new ArrayList<>();
+
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
       // If the event is longer than a day, no time slots are available to book a meeting.
-      return times; 
+      return new ArrayList<>();
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Sorry, I take back what I said. I think it makes more sense to return timesForEveryone as you are also setting this value.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -22,59 +22,87 @@
...
+  }
+
+  /**
+   * Adds a TimeRange to indicate a time gap for either {@code timesForEveryone} or {@code timesForConfirmed} depending
+   * on whether optional or confirmed attendees are attending the current {@code event}.
+   * 
+   * <p>Checks {@code request.duration} and {@code eventAttendees} to see if a time gap is possible for
+   * everybody, only confirmed attendees, only optional attendees, or no one at all.
+   */
+  public void setTimeGapIfPossible(MeetingRequest request, Set<String> eventAttendees, Event event) {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Should this method be private?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -57,7 +60,7 @@ public void setUp() {
   }
 
   @Test
-  public void optionsForNoAttendees() {
+  public void query_allDayAvailable_optionsForNoAttendees() {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: I am not sure I know what optionsForNoAttendees means if allDayAvailable?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -262,13 +260,156 @@ public void notEnoughRoom() {
...
 
     Collection<TimeRange> actual = query.query(events, request);
     Collection<TimeRange> expected = Arrays.asList();
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void query_ThreeAvailableSlots_optionalAttendeeWithAllDayEvent() {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]:
```suggestion
  public void query_threeAvailableSlots_optionalAttendeeWithAllDayEvent() {
```

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -262,13 +260,156 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void query_TwoAvailableSlots_optionalAttendeeWithEventBetweenTimeGap() {
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]:
```suggestion
  public void query_twoAvailableSlots_optionalAttendeeWithEventBetweenTimeGap() {
```

**Review by billie-cat — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 14 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Thanks for catching that! 
Done

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 13 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: done!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: done!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: gotcha, removed!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: gotcha, updated.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: yes, thanks for catching that. Updated!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 12 above)
Comment (amuamushu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: For this test, if there no attendees, the entire day is available for booking
I renamed the test methods such that the outcome (`allDayAvailable`) is at the end of the method name so hopefully, it is clearer. --> `query_noAttendees_allDayAvailable()`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -22,59 +22,86 @@
 
 /** Finds potential meeting times. */
 public final class FindMeetingQuery {
+  public Collection<TimeRange> timesForEveryone;
+  public Collection<TimeRange> timesForConfirmed;
+  public int currentEndTime;
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: class members should never be public. If you need to access them then you should create methods to return their values.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -22,59 +22,86 @@
...
-   * Returns available TimeRanges for the {@code request} given {@code events}, a Collection of
-   * all events so far. 
+   * Returns available TimeRanges for the {@code request} given {@code events}. 
    */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    Collection<TimeRange> times = new ArrayList<>();
-    Collection<String> attendees = request.getAttendees();
-    
+    timesForEveryone = new ArrayList<>();
+    timesForConfirmed = new ArrayList<>();
```
Comment (billie-cat) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: I think it made more sense when these used to be local variables and not class members. Your method is doing a lot. The JavaDoc claims it returns a TimeRange, but then it also sets all this internal state. Maybe the way it is implemented needs to be re-thought or this method needs to be split into multiple methods.

**Review by billie-cat — CHANGES_REQUESTED** [review]

**Review by billie-cat — APPROVED** [review]
